yuniql-platforms/sqlserver/SqlServerBulkImportService.cs

```diff
@@ -1,139 +1,249 @@
-                sqlBulkCopy.BulkCopyTimeout = commandTimeout.HasValue ? commandTimeout.Value : DEFAULT_CONSTANTS.COMMAND_TIMEOUT_SECS;
-                sqlBulkCopy.BatchSize = bulkBatchSize.HasValue ? bulkBatchSize.Value : DEFAULT_CONSTANTS.BULK_BATCH_SIZE;
-                sqlBulkCopy.EnableStreaming = true;
-                sqlBulkCopy.SqlRowsCopied += SqlBulkCopy_SqlRowsCopied;
-                foreach (var column in dataTable.Columns)
-                {
-                    sqlBulkCopy.ColumnMappings.Add(column.ToString(), column.ToString());
-                }
-                sqlBulkCopy.WriteToServer(dataTable);
-            }
-        }
-
-        private void SqlBulkCopy_SqlRowsCopied(object sender, SqlRowsCopiedEventArgs e)
-        {
-            _traceService.Info($"SqlServerBulkImportService copied {e.RowsCopied} rows");
-        }
-    }
-}
-
+
+        }
+
+        internal class DbTypeMap
+        {
+            public string ColumnName { get; set; }
+            public string SqlTypeName { get; set; }
+            public Type DotnetType { get; set; }
+        }
+
+        private Dictionary<string, DbTypeMap> LookupColumnDataTypes(string tableName, string schemaName, SqlConnection connection, SqlTransaction transaction)
+        {
+            var types = new Dictionary<string, DbTypeMap>();
+            string sql = $"SELECT COLUMN_NAME, DATA_TYPE FROM INFORMATION_SCHEMA.columns c where c.TABLE_NAME = '{tableName}'";
+
+            using (SqlCommand command = new SqlCommand(sql, connection, transaction))
+            {
+                using SqlDataReader reader = command.ExecuteReader();
+
+                while (reader.Read())
+                {
+                    var dbTypeMap = new DbTypeMap
+                    {
+                        ColumnName = reader.GetString(0),
+                        SqlTypeName = reader.GetString(1)
+                    };
+
+                    var dotnetType = MapSqlTypeToNative(dbTypeMap.SqlTypeName);
+
+                    if (dotnetType == null)
+                    {
+                        //not supported types: xml, rowversion, sql_variant, image, varbinary(max), binary, varbinary, timestamp
+                        throw new NotSupportedException($"SqlServerBulkImportService: Data type '{dbTypeMap.SqlTypeName}' on destination table {schemaName}.{tableName} is not support for bulk import operations.");
+                    }
+                    else
+                    {
+                        dbTypeMap.DotnetType = dotnetType;
+                        types.Add(dbTypeMap.ColumnName, dbTypeMap);
+                    }
+                }
+            }
+
+            return types;
+        }
+
+        private Type MapSqlTypeToNative(string dataType)
+        {
+            if (dataType == "char" || dataType == "nchar" || dataType == "text" || dataType == "ntext" || dataType == "varchar" || dataType == "nvarchar")

```

Comment:
Curious for the reasoning behind using a bunch of if/else if statements instead of a nice digestible switch statement

Reply:
In this project it is primarily for keeping the same look-and-feel as `PostgreSqlBulkImportService.cs`. In general I also find this code easy to read. It may not be best practice and it may be a tad slower, but I still find it nice and readable.

Reply:
Looking around a bit, I do see that now. Thanks for the info!

Reply:
@RobertWilc Thanks for looking out the code and @automada  was right in the consistency/readability reasoning.

@automada In other implementations, we named this method `GetDestinationSchema`. A new connection is also initated here instead of passing existing connection. Can't remember why I did it this way but perhaps to isolate the connection if the bulk upload is happenign in streaming fashion, avoid uninintendted side effect.

I will akso mark this `GetDestinationSchema` and `ParseCsv` as something we can generalize to a base class, extension method or another class we can inject. It's duplicared several times already :)